### PR TITLE
Extended online scaler. It is now possible to configure type of conce…

### PR DESCRIPTION
**Pull request by alek-thunder**
…pt drift detection. Besides threshold-based, it allows an threshold-based detection on a windwo of values in order to become robust against outlier and measurement errors.

### CODE REVIEW

File: src/main/java/bitflow4j/steps/normalization/OnlineAutoMinMaxScaler.java [hunk 1]
```diff
@@ -172,4 +177,57 @@ public String toString() {
...
+
+        @Override
+        public boolean isConceptChanged(double value) {
+            double referenceValue = value;
+            if (threshold > 0) {
+                values.add(value);
+                if (values.size() >= windowSize) {
+                    int index = Math.round((float) windowSize / 2);
+                    if (index >= 0 && index < values.size()) {
+                        referenceValue = values.get(index);
```
Comment (antongulenko) [inline src/main/java/bitflow4j/steps/normalization/OnlineAutoMinMaxScaler.java:222]: I'm not sure I understand the logic behind this... Shouldn't this compute the average of the window or something? I I understand correctly, this should make everything the concept-changing robust to spikes, but this way the spike might randomly be the one selected comparsion value.

File: src/main/java/bitflow4j/steps/normalization/AbstractOnlineScaler.java [hunk 2]
```diff
@@ -25,6 +25,22 @@
         List<Boolean> conceptChanged(AbstractOnlineScaler scaler, List<String> changedFeatures);
     }
 
+    public interface ConceptChangeDetector {
+        boolean isConceptChanged(double value);
+
+        static ConceptChangeDetector getInstance(String type, double threshold) {
```
Comment (antongulenko) [inline src/main/java/bitflow4j/steps/normalization/AbstractOnlineScaler.java:31]: Where is this method used? (sorry, did not check in IDE).

**Review by antongulenko — CHANGES_REQUESTED** [review]

File: src/main/java/bitflow4j/steps/normalization/OnlineAutoMinMaxScaler.java (hunk 1 above)
Comment (antongulenko) [reply src/main/java/bitflow4j/steps/normalization/OnlineAutoMinMaxScaler.java:222]: Probably, the values list was meant to be sorted, before picking the value from the middle?... I don't want to make the fix myself before checking with @alek-thunder

Comment (alek-thunder) [reply src/main/java/bitflow4j/steps/normalization/OnlineAutoMinMaxScaler.java:222]: Very good point. Will fix this. As your 2nd comment correctly states, my idea is to sort the window before accessing the middle element. Therefore sparse outliers (also in both directions for min max scaling) are ignored for determining the new min and max values that should be used for scaling. Already pushed a fix.

File: src/main/java/bitflow4j/steps/normalization/AbstractOnlineScaler.java (hunk 2 above)
Comment (alek-thunder) [reply src/main/java/bitflow4j/steps/normalization/AbstractOnlineScaler.java:31]: OnlineAutoMinMaxScaler and ScaleMinMaxWithModelRepository

**Review by antongulenko — APPROVED** [review]

**Pull request merged**